### DISCUSSION TITLE
A very expensive three phase meter?

### DISCUSSION THREAD

**Discussion by BuongiornoTexas** [jasonacox/Powerwall-Dashboard#83] [Polls]
I've been puttering around with pypowerwall and Powerwall-Dashboard, and I've realised that it could be used as a rough per phase power meter for anyone with a three phase/3 phase system. I think with a bit of work, I can get a per phase plot of grid, home/site load, solar and battery.  This will allow monitoring of the three phase net metering, and could also be useful when trying to assign circuits to phases and when troubleshooting misbehaving circuits.

I'm definitely going to do a hack first pass of this for fun and personal interest. If other people think this feature may be useful, I will spend a bit more time to put a PR together to automate the three phase monitoring setup.

I've put together a proof of concept python script which generates output along the following lines (at the end of this LOOOOONG post):

```
Grid = -4542 (check = -4566)
Solar = 8723 (check = 8733)
Home load = 394.5 (check = 447.0)
Battery = -3730 (check = -3720.0)

Phase a: grid = -900, solar = 5020, home load = 400.0, battery = -3720.0
Phase b: grid = 0, solar = 0, home load = 0.0, battery = 0.0
Phase c: grid = -3666, solar = 3713, home load = 47.0, battery = 0.0
```

The first block shows aggregate meter values and cross checks from the sum of the phase values. The second block shows per phase values. Note that the data will not provide exact values, as I'm pulling some cached data (time drift), I need to make 4 different api calls to get the full data set (more time drift), and there is a bit of variation between the different cts and meters - I'm seeing variations of anywhere between 10 and 100w  on the reconciliations (e.g. phase c has a consistent errror between the solar meter and the gateway site meter). But this is good enough for what I want to do. 

So, could I get poll responses on level of interest? And could I also ask anybody with a three phase system to run the proof of concept script and let us know the outcome to debug/improve the phase data extraction (our system is pretty simple - one power wall on one phase only - data from systems with more powerwalls on single and multiple phases will be very useful). **Note**: You need to enter your local powerwall login credentials to run the script!

threephasetest.py
```
from collections import defaultdict
from tokenize import Double
from typing import DefaultDict
import pypowerwall

# Optional: Turn on Debug Mode
# pypowerwall.set_debug(True)

# Credentials for your Powerwall - Customer Login Data
password='qqqqqq'
email='xxxx.yyyy@aaa.bbb'
host = "192.168.1.xxx"         # Address of your Powerwall Gateway
timezone = "Australia/Sydney"  # Your local timezone

# Connect to Powerwall
pw = pypowerwall.Powerwall(host,password,email,timezone)

# Some System Info
print("Site Name: %s - Firmware: %s - DIN: %s" % (pw.site_name(), pw.version(), pw.din()))
print("System Uptime: %s\n" % pw.uptime())

# Looks as though tesla uses both phase 1, 2, 3 and a, b, c
# I'm making the bold assumption that Phase 1 = Phase a, etc. as this matches up 
# with our installation. Can adapt if others have different experiences. Will work
# with abc as the default, as this seems to be used more often 
phases = ["a", "b", "c"]

# Find powerwalls and link to Phase. May be a better way to do this, but it is a start
pwjson = pw.poll('/api/powerwalls', jsonformat = True)["powerwalls"]
batterystatusjson = pw.poll('/api/system_status', jsonformat = True)["battery_blocks"]
# If I was being really paranoid, I'd check both part number and SN, 
# but really, how many pws will a residential install have - low prob of collisions
# on serial number only, so ignore this frill.
# Do a check that serials are consistent in powerwalls and system_status
for pwall, bat in zip(pwjson, batterystatusjson):
    sn1 = pwall["PackageSerialNumber"] 
    sn2 = bat["PackageSerialNumber"]
    if sn1 != sn2:
        raise ValueError(f"Mismatched powerwall serials {sn1} and {sn2}")  

# map pws/batteries to phases
pwphasemap = [ phases[int(pw["phase"][-1])-1] for pw in pwjson ]

# grab data in sequential calls to minimise drift in values
aggregates = pw.poll('/api/meters/aggregates', jsonformat = True)
solar_j = pw.poll('/api/meters/solar', jsonformat = True)[0]["Cached_readings"]
grid_j =  pw.poll('/api/meters/site', jsonformat = True)[0]["Cached_readings"]
batterystatusjson = pw.poll('/api/system_status', jsonformat = True)["battery_blocks"]

grid_t = aggregates["site"]["instant_power"]
solar_t = aggregates["solar"]["instant_power"]
home_t = aggregates["load"]["instant_power"]
battery_t = aggregates["battery"]["instant_power"]

battery_p: DefaultDict[str, float] = defaultdict(float)
for bat, ph_id in zip(batterystatusjson, pwphasemap):
    battery_p[ph_id] += bat["p_out"]

solar_p = { ph_id: solar_j.get("real_power_"+ph_id, 0) for ph_id in phases}
grid_p = { ph_id: grid_j.get("real_power_"+ph_id, 0) for ph_id in phases}
home_p = { ph_id: solar_p[ph_id] + grid_p[ph_id] + battery_p[ph_id] for ph_id in phases}

print()

print (f"Grid = {grid_t} (check = {sum(grid_p.values())})")
print(f"Solar = {solar_t} (check = {sum(solar_p.values())})")
print(f"Home load = {home_t} (check = {sum(home_p.values())})")
print(f"Battery = {battery_t} (check = {sum(battery_p.values())})")

print()

for ph_id in phases:
    print(f'Phase {ph_id}: grid = {grid_p[ph_id]}, solar = {solar_p[ph_id]}, '
       f'home load = {home_p[ph_id]}, battery = {battery_p[ph_id]}')
```

**Comment by OzRocky** [comment]
Whoopsie. That should have been "Yes! But I don't have a three phase system (yet)."

**Reply by BuongiornoTexas** [reply]
> Whoopsie. 

No worries mate - preliminary implementation is in https://github.com/jasonacox/Powerwall-Dashboard/discussions/94#discussion-4439911. I'm still working on it, and will continue to update there as I make progress.
